### PR TITLE
Fix bug in AstroCalc/Eclipses for local solar eclipses

### PR DESCRIPTION
**Pull request by worachate001**
### Description
This bug in AstroCalc/Eclipses is found when total/annular eclipses take place very close to the horizon.

**To reproduce**
- Set the location to North Pole
- Open 'AstroCalc/Local Solar Eclipses' tab
- Set 2015 as the starting year

The calculation shows that 2015 Mar 20 was a partial eclipse, contradicting to the magnitude at greatest eclipse of 1.013 (total eclipse). The bug happens because the total eclipse took place very close to the horizon. The same situation can be found on other dates when your location is at the start or end point of any central eclipses.

This patch will also avoid to show time of 1st contact or last contact when total or annular eclipses are in progress at Sunrise/Sunset.

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] This change requires a documentation update

### How Has This Been Tested?

Test cases:
Lat. 26.0
Long. 47.5
Date: 2019-12-26 (annular eclipse at Sunrise)

Lat. 47.4
Long. -19.6
Date: 2024-04-08 (total eclipse at Sunset)

**Test Configuration**:
* Operating system: Windows 10

## Checklist:

- [x] My code follows the [code style](http://stellarium.org/doc/head/codingStyle.html) of this project.
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (header file)
- [ ] I have updated the respective chapter in the Stellarium User Guide
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/gui/AstroCalcDialog.cpp [hunk 1]
```diff
@@ -3539,6 +3539,9 @@ void AstroCalcDialog::generateSolarEclipsesLocal()
 							treeItem->setData(SolarEclipseLocalDate, Qt::UserRole, JDmax);
 							treeItem->setText(SolarEclipseLocalType, eclipseTypeStr);
 							treeItem->setText(SolarEclipseLocalFirstContact, QString("%1").arg(localeMgr->getPrintableTimeLocal(JD1)));
+							if (centraleclipse)
```
Comment (alex-w) [inline src/gui/AstroCalcDialog.cpp:3542]: Why not use `if (centraleclipse && JD2<JD1)`?

File: src/gui/AstroCalcDialog.cpp [hunk 2]
```diff
@@ -3555,6 +3558,9 @@ void AstroCalcDialog::generateSolarEclipsesLocal()
 								treeItem->setText(SolarEclipseLocal3rdContact, dash);
 							treeItem->setToolTip(SolarEclipseLocal3rdContact, q_("The time of third contact"));
 							treeItem->setText(SolarEclipseLocalLastContact, QString("%1").arg(localeMgr->getPrintableTimeLocal(JD4)));
+							if (centraleclipse)
```
Comment (alex-w) [inline src/gui/AstroCalcDialog.cpp:3561]: Why not use `if (centraleclipse && JD3>JD4)`?

File: src/gui/AstroCalcDialog.cpp (hunk 1 above)
Comment (worachate001) [reply src/gui/AstroCalcDialog.cpp:3542]: Yes, I'm about to change that. Thanks.

File: src/gui/AstroCalcDialog.cpp (hunk 2 above)
Comment (worachate001) [reply src/gui/AstroCalcDialog.cpp]: Thanks, I have added else... after that. Decided to remove it and forgot to change it.

**Comment by worachate001** [comment]
I would like to note that situations like this are sometimes difficult to handle. Not showing first or last contact maybe a bit weird. But there are cases like this when totality or annularity begins just before Sunrise and ends just after that. So first contact actually occurs but we can't see it.

**Review by alex-w — APPROVED** [review]
Thanks!

**Pull request merged**